src/object_gui.cpp

```diff
@@ -544,8 +544,37 @@ class BuildObjectWindow : public Window {
 	void OnPlaceObject(Point pt, TileIndex tile) override
 	{
 		ObjectClass *objclass = ObjectClass::Get(_selected_object_class);
-		Command<CMD_BUILD_OBJECT>::Post(STR_ERROR_CAN_T_BUILD_OBJECT, CcPlaySound_CONSTRUCTION_OTHER,
+		if (_settings_game.construction.build_object_area_permitted && objclass->GetSpec(_selected_object_index)->size == 0x11) {
+			VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_BUILD_OBJECT);
+		} else {
+			Command<CMD_BUILD_OBJECT>::Post(STR_ERROR_CAN_T_BUILD_OBJECT, CcPlaySound_CONSTRUCTION_OTHER,
 				tile, objclass->GetSpec(_selected_object_index)->Index(), _selected_object_view);
+		}
+	}
+
+	void OnPlaceDrag(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt) override
+	{
+		VpSelectTilesWithMethod(pt.x, pt.y, select_method);
+	}
+
+	void OnPlaceMouseUp(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt, TileIndex start_tile, TileIndex end_tile) override
+	{
+		if (pt.x != -1) {
+			switch (select_proc) {
+			default: NOT_REACHED();
+			case DDSP_BUILD_OBJECT:
+				if (!_settings_game.construction.freeform_edges) {
+					/* When end_tile is MP_VOID, the error tile will not be visible to the
+					 * user. This happens when terraforming at the southern border. */
+					if (TileX(end_tile) == MapMaxX()) end_tile += TileDiffXY(-1, 0);
+					if (TileY(end_tile) == MapMaxY()) end_tile += TileDiffXY(0, -1);
+				}
+				ObjectClass *objclass = ObjectClass::Get(_selected_object_class);
+				Command<CMD_BUILD_OBJECT>::Post(STR_ERROR_CAN_T_BUILD_OBJECT, CcPlaySound_CONSTRUCTION_OTHER,
```

Comment:
If you look at the original commit at JGRPP, in the `OnPlaceMouseUp` function the `DoCommandP` called with `CMD_BUILD_OBJECT_AREA`. So I think you probably also should called the `Command` here with `CMD_BUILD_OBJECT_AREA`.

```suggestion
				Command<CMD_BUILD_OBJECT_AREA>::Post(STR_ERROR_CAN_T_BUILD_OBJECT, CcPlaySound_CONSTRUCTION_OTHER,
```